woocommerce/woocommerce
# E2E frontend testing

This is not necessarily the finished implementation, but is an implementation of an approach I think will work nicely. I've created this pull-request because it is nearing completion, and this is a good point to go over it and change things if necessary. I wouldn't necessarily merge this into master right this second. :)

**The e2e testing has two parts**: The page components and the tests.

**The components** are a library that simplifies interaction with pages to make it simpler to write tests and enables us to easily alter the behind-the-scenes implementations of things while not having to rewrite all of the tests. This is similar to the WooCommerce CRUD, where we can modify the data store to do whatever behind-the-scenes and, as long as the methods are the same, everything should just work. The components have methods like ['clickAddProducts' and 'setRefundAmount'](https://github.com/woocommerce/wc-e2e-page-objects/blob/master/src/components/wp-admin/component-meta-box-order-items.js). The components live in the wc-e2e-page-objects repo and are available through npm. 

Having the components available in their own npm repo makes it easy for extension developers to use the library in their own testing, without requiring the WooCommerce source code to use them.

**The tests** use the components to verify the presence of, ability to interact with, and correct output of elements around the site. The tests do this by using the components to navigate and interact with the site. The tests live in the WooCommerce repo, so that contributors familiar with local WooCommerce development can easily add, modify, and run tests.

**To run the tests:**

1. Set up a clean WooCommerce install and add a basic flat-rate-shipping zone (The shipping address checkbox on the checkout page needs to be available for some of the tests). Any standard theme should work (I've tested on 2017 and 2012).

2. Import the dummy data at https://github.com/woocommerce/woocommerce/blob/master/dummy-data/dummy-data.xml or in your WooCommerce folder.

3. `npm install` to get all of the dependencies.

4. Update your site/creds in `tests/frontend-tests/config/development.json`.

5. `grunt frontend-tests`

6. Sit back and sip your coffee/tea/Coke.

Can someone unfamiliar with this give this a shot and verify the setup and running of the tests is easy? @mikejolley, @claudiosanches, anyone else reading this pull request? :)

@gedex do you see any problems with this approach? The main thing I see that we'd have to work out is whether to duplicate the tests in both the WooCommerce and wc-e2e-page-objects repo?

------

TODO: 
- Remove config files and make config come from command line
- Add chromedriver to package.json
- Aadjust the engines section of package.json/specify a correct minimum version. Able to run the tests twice successfully using 6.10.0 (latest) but not 6.9.4.
